docs/enhancements/infrastructure-monitoring-service-api/alarms.md

```diff
@@ -0,0 +1,669 @@
+
+-- probable_cause will be auto populated
+```
+Notes on Init phase 
+- `versions` table reflects unique `major-minor` version `ManagedClusters` currently deployed.
+  To get available `major-minor` managed cluster, we can list `ManagedCluster` CR and look for label `openshiftVersion-major-minor`.
+    ```shell
+    oc get managedclusters
+    ```
+  
+    ```yaml
+    apiVersion: cluster.open-cluster-management.io/v1
+    kind: ManagedCluster
+    metadata:
+      labels:
+        openshiftVersion-major-minor: "4.16"
+    ```
+
+- `alarm_dictionary` essentially links `ResourceTypeID` and `Version`. The conversion can be seen below.
+
+    | managed_cluster<br/>`from alerts`<br/> | resourceID  <br/>`same as managed_cluster`<br/> | resourceTypeID for caas<br/>`derived from (resourceID + ResourceKindLogical + ResourceClassCompute)`<br/> |
+    |----------------------------------------|-------------------------------------------------|-----------------------------------------------------------------------------------------------------------|
+    | f90561e2-6420-4924-b081-f4f8eaf50618   | f90561e2-6420-4924-b081-f4f8eaf50618            | 4586f964-6c6f-407b-9b18-cb3c9a712ec4                                                                      |
+   
+    `entity_type` data should be coming from Inventory API but for now we can hard-code it. `telco-model-OpenShift-<Full Version>`
+- `alarm_definitions` reflects Rules in PromRule CR. We only grab the full set based on unique entries in `Versions` table 
+   
+   ```yaml
+    # Partial PrometheusRule to show NodeClockNotSynchronising
+    apiVersion: monitoring.coreos.com/v1
+    kind: PrometheusRule
+    metadata:
+      name: node-exporter-rules
+      namespace: openshift-monitoring
+    spec:
+      groups:
+        - name: node-exporter
+          rules:
+            - alert: NodeClockNotSynchronising  #### (alarm_definitions.alarm_name)
+              annotations:
+                description: Clock at {{ $labels.instance }} is not synchronising. Ensure NTP is configured on this host.
+                runbook_url: https://github.com/openshift/runbooks/blob/master/alerts/cluster-monitoring-operator/NodeClockNotSynchronising.md #### (alarm_definitions.proposed_repair_actions)
+                summary: Clock not synchronising. #### (alarm_definitions.alarm_description)
+              expr: |
+                min_over_time(node_timex_sync_status{job="node-exporter"}[5m]) == 0
+                and
+                node_timex_maxerror_seconds{job="node-exporter"} >= 16
+              for: 10m
+              labels: 
+                severity: critical #### (alarm_definitions.extensions)
+    ```
+    - Use ACM to get credentials of the unique major.minor clusters and retrieve all the PrometheusRules from them to parse. 
```

Comment:
Do PrometheusRules ever change or new ones get added (e.g. on a z-stream upgrade)? Or if a cluster gets installed with a release that wasn't being managed yet? What will be the strategy to handle that?

Reply:
> ever change or new ones get added

Yeah potentially? But at every restart we will audit the managedCluster (added this bit in the doc) and if we pick the latest one (i.e between 4.14.4 and 4.14.6 we pick 4.14.6) it should be mitigate some of concerns.

> release that wasn't being managed yet
We can't do that. If a cluster is not managed, hub cant send us the alerts.

Reply:
I was thinking of the case where a hub had only 4.16 clusters installed (for example) and the the first 4.18 cluster is installed. What would trigger the alarm service to read the 4.18 alarm definitions in that case?

Reply:
> What would trigger the alarm service to read the 4.18 alarm definitions 

Currently nothing tbh. So the only way to get the latest data (e.g new 4.18 cluster) we would be need to restart. But to get this info without a full restart we can need to setup some kind of "watch" over the managedcluster CRs. It could be done with the operator (so operator would requeue when it detects a change) and simply delete the alarms pod OR alarms server would directly watch the CRs and update DB as needed. But ultimately though...this could be something Inventory/resource server should alert us with("btw new cluster/resourceType added...update your DB")?

Maybe @browsell can weigh in on this.

Reply:
Could you just read the alarm definitions for all supported versions?

Reply:
i dont think that's possible since the available clusters tells us which definitions to store. 

We are retrieving PromRules and then generating alarm definitions (and finally storing it) based on the currently set of major.minor cluster as indicated by hub. In other words...we need first know that there's a new 4.18 cluster before actually retrieving and storing the definitions. And currently the only way to know this by auditing at startup